File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/chat/MaintenanceCommand.java
package net.bestia.zoneserver.chat;

import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import net.bestia.model.dao.AccountDAO;
import net.bestia.model.domain.Account;
import net.bestia.model.domain.Account.UserLevel;
import net.bestia.zoneserver.actor.ZoneAkkaApi;
import net.bestia.zoneserver.configuration.MaintenanceLevel;
import net.bestia.zoneserver.configuration.RuntimeConfigurationService;
import net.bestia.zoneserver.service.LoginService;

/**
 * Allows admins to set the server into maintenance mode. Use with caution since
 * this will disconnect ALL users! It can only switch the server into partial
 * maintenance since full maintainance would also disconnect the admin itself.
 * 
 * @author Thomas Felix
 *
 */
@Component
public class MaintenanceCommand extends BaseChatCommand {

	private static final Logger LOG = LoggerFactory.getLogger(MaintenanceCommand.class);

	private static final String CMD_START_REGEX = "^/maintenance .*";
	private static final Pattern CMD_PATTERN = Pattern.compile("/maintenance (.+)");

	private final RuntimeConfigurationService config;
	private final LoginService loginService;


	@Autowired
	public MaintenanceCommand(
			AccountDAO accDao, 
			ZoneAkkaApi akkaApi, 
			LoginService loginService, 
			RuntimeConfigurationService config) {
		super(accDao, akkaApi);

		this.loginService = Objects.requireNonNull(loginService);
		this.config = Objects.requireNonNull(config);
	}

	@Override
	public boolean isCommand(String text) {
		return text.matches(CMD_START_REGEX);
	}

	@Override
	public UserLevel requiredUserLevel() {
		return UserLevel.ADMIN;
	}

	@Override
	protected void executeCommand(Account account, String text) {

		// Extract name of the new map.
		final Matcher match = CMD_PATTERN.matcher(text);

		if (!match.find()) {
			printError(account.getId());
			return;
		}

		boolean isMaintenance;

		try {
			final String txt = match.group(1);
			isMaintenance = Boolean.parseBoolean(txt);
		} catch (Exception e) {
			printError(account.getId());
			return;
		}
		
		LOG.info("Account {} set maintenance to: {}", account.getId(), isMaintenance);

		if(isMaintenance) {
			sendSystemMessage(account.getId(), "Server maintenance: true");
			config.setMaintenanceMode(MaintenanceLevel.PARTIAL);
			loginService.logoutAllUsersBelow(UserLevel.SUPER_GM);
		} else {
			sendSystemMessage(account.getId(), "Server maintenance: false");
			config.setMaintenanceMode(MaintenanceLevel.NONE);
		}
	}

	private void printError(long accId) {
		sendSystemMessage(accId, "Usage: /maintenance [true, false]");
	}
}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/chat/MaintenanceCommand.kt
package net.bestia.zoneserver.chat

import mu.KotlinLogging
import net.bestia.model.account.Account
import net.bestia.model.account.AccountType
import net.bestia.model.server.MaintenanceLevel
import net.bestia.zoneserver.actor.MessageApi
import net.bestia.zoneserver.account.LogoutService
import net.bestia.zoneserver.config.RuntimeConfigService
import org.springframework.stereotype.Component

private val LOG = KotlinLogging.logger { }

/**
 * Allows admins to set the server into maintenance mode. Use with caution since
 * this will disconnect ALL users! It can only switch the server into partial
 * maintenance since full maintainance would also disconnect the admin itself.
 *
 * @author Thomas Felix
 */
@Component
internal class MaintenanceCommand(
    messageApi: MessageApi,
    private val logoutService: LogoutService,
    private val configService: RuntimeConfigService
) : BaseChatCommand(messageApi) {

  override val helpText: String
    get() = "Usage: /maintenance [TRUE|FALSE]"

  override fun isCommand(text: String): Boolean {
    return text.matches(CMD_START_REGEX)
  }

  override fun requiredUserLevel(): AccountType {
    return AccountType.ADMIN
  }

  override fun executeCommand(account: Account, text: String) {
    val match = CMD_PATTERN.find(text)

    if (match == null) {
      printError(account.id)
      return
    }

    val isMaintenance = match.groups[1]?.value?.toBoolean() ?: false
    LOG.info("Account {} set maintenance to: {}", account.id, isMaintenance)

    if (isMaintenance) {
      sendSystemMessage(account.id, "Server maintenance: true")
      configService.runtimeConfig = configService.runtimeConfig
          .copy(maintenanceLevel = MaintenanceLevel.PARTIAL)
      logoutService.logoutAllUsersBelow(AccountType.SUPER_GM)
    } else {
      sendSystemMessage(account.id, "Server maintenance: false")
      configService.runtimeConfig = configService.runtimeConfig
          .copy(maintenanceLevel = MaintenanceLevel.NONE)
    }
  }

  private fun printError(accId: Long) {
    sendSystemMessage(accId, "Usage: /maintenance [true, false]")
  }

  companion object {
    private val CMD_START_REGEX = "^/maintenance .*".toRegex()
    private val CMD_PATTERN = "/maintenance (true|false)".toRegex(RegexOption.IGNORE_CASE)
  }
}
